### DISCUSSION TITLE
Remote server returned an error: (403) Forbidden.

### DISCUSSION THREAD

**Discussion by susrutpatnaik15** [microsoft/PowerShellForGitHub#375] [Q&A]
I'm trying to automate GitHub Pull request via this PS script:

```powershell
$apisUrl = "$($env:SYSTEM_TEAMFOUNDATIONCOLLECTIONURI)/$($env:SYSTEM_TEAMPROJECT)/_apis"#
$projectUrl = "https://github.com/user/**********.git"
$user = "github_user"
$token = "github_pat_token"
# Base64-encodes the Personal Access Token (PAT) appropriately#
$base64AuthInfo = [Convert]::ToBase64String([Text.Encoding]::ASCII.GetBytes(("{0}:{1}" -f $user,$token)))

# create common headers
$headers = @{}
#$headers.Add("Authorization", "Bearer $env:SYSTEM_ACCESSTOKEN")
$headers.Add('Authorization',('Basic {0}' -f $base64AuthInfo))
$headers.Add('Accept','application/json')
$headers.Add("Content-Type", "application/json")

# Specify HTTP method
$method = "get"

[Net.ServicePointManager]::SecurityProtocol = "tls12, tls11, tls"

# Create a Pull Request
$pullRequestUrl = "$projectUrl/pullrequests?api-version=5.1"
$pullRequest = @{
        "sourceRefName" = "refs/heads/repo1Path2"
        "targetRefName" = "refs/heads/repo1Path3"
        "title" = "Pull from repo1Path2 to repo1Path3"
        "description" = ""
    }

$pullRequestJson = ($pullRequest | ConvertTo-Json -Depth 5)

Write-Output "Sending a REST call to create a new pull request from repo1Path2 to repo1Path3"

# REST call to create a Pull Request
#$pullRequestResult = Invoke-RestMethod -Method POST -Headers $headers -Body $pullRequestJson -Uri $pullRequestUrl;
$pullRequestResult = Invoke-RestMethod -Method POST -Headers @{Authorization=("Basic {0}" -f $base64AuthInfo)} -Body $pullRequestJson -Uri $pullRequestUrl;
$pullRequestId = $pullRequestResult.pullRequestId

Write-Output "Pull request created. Pull Request Id: $pullRequestId"

# Set PR to auto-complete
$setAutoComplete = @{
    "autoCompleteSetBy" = @{
        "id" = $pullRequestResult.createdBy.id
    }
    "completionOptions" = @{       
        "deleteSourceBranch" = $False
        "bypassPolicy" = $False
    }
}

$setAutoCompleteJson = ($setAutoComplete | ConvertTo-Json -Depth 5)

Write-Output "Sending a REST call to set auto-complete on the newly created pull request"

# REST call to set auto-complete on Pull Request
$pullRequestUpdateUrl = ($projectUrl + '/pullRequests/' + $pullRequestId + '?api-version=5.1')

$setAutoCompleteResult = Invoke-RestMethod -Method PATCH -Headers $headers -Body $setAutoCompleteJson -Uri $pullRequestUpdateUrl**

Write-Output "Pull request set to auto-complete"
```
-----------
But I'm fetching this error 
```
Invoke-RestMethod : The remote server returned an error: (403) Forbidden.
At C:\Users\*******\*******\GHPR_PSScript6_Ado.ps1:58 char:26
+ ... eteResult = Invoke-RestMethod -Method PATCH -Headers $headers -Body $ ...
+                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (System.Net.HttpWebRequest:HttpWebRequest) [Invoke-RestMethod], WebException
    + FullyQualifiedErrorId : WebCmdletWebResponseException,Microsoft.PowerShell.Commands.InvokeRestMethodCommand**
```

Please check if the script has any issues, if not how to authenticate. Thanks in advance.

> Note: Post edited to improve formatting

**Comment by HowardWolosky** [comment]
Hi @susrutpatnaik15,

I do not have the time at the moment to fully debug your posted script, in large part because you're not attempting to even use this module at all.

Most of what you're doing can be achieved with this module in a much simpler manner...

You should review [`Get-GitHubPullReques`](https://github.com/microsoft/PowerShellForGitHub/wiki/Get-GitHubPullRequest) and [`New-GitHubPullRequest`](https://github.com/microsoft/PowerShellForGitHub/wiki/New-GitHubPullRequest), and then you can use [`Invoke-GHRestMethod`](https://github.com/microsoft/PowerShellForGitHub/wiki/Invoke-GHRestMethod) to do the equivalent of `Set-GitHubPullRequest` which is not yet implemented in the module.

**Reply by HowardWolosky** [reply]
What I can tell you is that the URI that you are sending your raw REST requests to is incorrect.  You're sending it to the URI where your git project can be downloaded from as opposed to the GitHub API REST endpoint.

**Reply by susrutpatnaik15** [reply]
Hi @HowardWolosky,

I reviewed the modules ([Get-GitHubPullReques](https://github.com/microsoft/PowerShellForGitHub/wiki/Get-GitHubPullRequest) and [New-GitHubPullRequest](https://github.com/microsoft/PowerShellForGitHub/wiki/New-GitHubPullRequest)) as suggested but was unable to run the script with my set of param values in the provided script. I'm still exploring PowerShell, may be I'm not following it correctly. 

Regards
Susrut

**Reply by HowardWolosky** [reply]
The simplest version of what you're trying to do, but with this module:

First, be sure you've [_installed_ the module](https://github.com/microsoft/PowerShellForGitHub#installation).

Then:

```powershell
Set-GitHubAuthentication # enter anything in the username field and then your PAT as the password
$pr = New-GitHubPullRequest -OwnerName <project owner> -RepositoryName <projectName> -Title 'Pull from repo1Path2 to repo1Path3' -Head repo1Path2 -Base repo1Path3
```

The thing that's throwing me here is that you really seem to be wanting to do Azure DevOps stuff here.  There isn't the concept of "auto-complete" with GitHub as there is with ADO.  Have you looked at the API documentation for GitHub at all?  It's not clear to me at this point what you're actually trying to accomplish.

> [GitHub Pull Request API's](https://docs.github.com/en/rest/pulls/pulls)

**Reply by susrutpatnaik15** [reply]
Requirement: **Create a PowerShell script to automate Pull Requests & Merge of the release branches (Head & Base branch) in GitHub.**

As you have mentioned few steps, I had already installed the module & even set the authentication part. Moreover, I have also followed the API documentation for GitHub.

**Comment by susrutpatnaik15** [comment]
Thanks for your prompt response. I'll check the same.
